GPT Scenario Generation

ZhiLinR opened the pull request:
core functionalities for scenario generation complete
not dockerised/deployed yet (have to look into it i think)

import files are ended with js currently since the editor didn't recognise imports without a file extension (😔)

cre8tion commented on microsvcs/gpt_generation/README.md:
````diff
@@ -0,0 +1,85 @@
+
+Sample success response: 
+```
+{
````
Not sure if you storing these scenarios.

Can consider generating a corresponding scenario ID

cre8tion commented on microsvcs/gpt_generation/README.md:
````diff
@@ -0,0 +1,85 @@
+
+```
+---
+## POST /api/scenario/end
````
Can use PUT with the same route as above, with scenario id as part of the request body

ZhiLinR replied:
corrected

cre8tion commented on microsvcs/gpt_generation/README.md:
```diff
@@ -0,0 +1,85 @@
+## POST /api/scenario/end
+*naming improvements appreciated 😢* 
+
+Logs the user's interaction with the scenario (namely the given scenario and the selected option).
```
In this case, its more like choosing a selected option rather than logging

cre8tion commented on microsvcs/gpt_generation/README.md:
````diff
@@ -0,0 +1,85 @@
+```
+
+---
+## GET /api/scenario/summary/:uid
````
Looks fine

cre8tion commented on microsvcs/gpt_generation/app/src/controller/scenario_con.ts:
```diff
@@ -0,0 +1,45 @@
+
+// taking ideas on better names - this function is for when the user actually selects their decision
+// itll record their option and store it in the database
+scenarioRoute.post('/scenario/end', (req: Request, res: Response): void => {
```
Commented above

cre8tion reviewed (commented):
LGTM generally

cre8tion commented:
Any updates on this branch?

ZhiLinR commented:
not atm, i think i can implement the OpenAI API retrieval for the scenario generation with [Threads ](https://platform.openai.com/docs/api-reference/threads/object)and [Messages](https://platform.openai.com/docs/api-reference/messages/listMessages) 

will also implement the other comments by tmr

ZhiLinR commented:
base implementation of threads and message logic implemented in (works but could use some finetuning, theres some .then() chaining and function nesting)(spaghetti even). readme also includes one of the behavioral lapses in the endpoint.

ZhiLinR commented:
i think this is temporarily like the basic features for the scenario generation, though i think will open a new issue if theres time to explore implementations with Google Analytics and other interactions

cre8tion commented on microsvcs/gpt_generation/README.md:
````diff
@@ -0,0 +1,129 @@
+```
+
+### Additional Info:
+This API's current behaviour has a behavioural lapse for the microservice. Only this endpoint can create a new Thread
````
Why will this be a lapse

cre8tion commented on microsvcs/gpt_generation/README.md:
````diff
@@ -0,0 +1,129 @@
+```
+
+---
+## PUT /api/scenario/end
````
PUT /api/scenario is fine, unless you really intend this to be the final response for the scenario

cre8tion commented on microsvcs/gpt_generation/app/src/controller/profile_con.ts:
```diff
@@ -0,0 +1,26 @@
+        const uid: string = req.params.uid;
+        const scenario = new Scenario(uid);
+        const userThread = new OAThreadsAPI(uid)
+        UModel.getUserThread(uid).then((result) => {
```
Can consider exploring async/await in the future

cre8tion commented on microsvcs/gpt_generation/app/src/controller/profile_con.ts:
```diff
@@ -0,0 +1,26 @@
+import express, { Express, Request, Response } from "express";
+import { Scenario } from '../ext_api/ScenarioClass.js'
+import { OAThreadsAPI } from "../ext_api/helpers/OAThreadsAPIClass.js";
+import * as UModel from '../model/user_model.js'
```
Name as UserModel will be better

cre8tion commented on microsvcs/gpt_generation/app/src/ext_api/init/openai_conf.ts:
```diff
@@ -0,0 +1,16 @@
+export async function OAConnect(content: string) {
+    const completion = await openai.chat.completions.create({
+        messages: [{ role: "system", content: content }],
+        model: "gpt-3.5-turbo",
```
gpt-4.o might be cheaper and if you want to explore

cre8tion commented on microsvcs/gpt_generation/app/src/ext_api/helpers/PromptClass.ts:
```diff
@@ -0,0 +1,62 @@
+/**
+ * Contains the text prompts to be inserted into OpenAI API.
+ * Lowkey what is this code :wtf:
```
How did you design this text btw

cre8tion commented on microsvcs/gpt_generation/app/src/ext_api/ScenarioClass.ts:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Extends {@link _Prompt} which allows for the text prompts to be accessed directly
+ */
+export class Scenario extends _Prompt {
```
Interesting use of class.....

cre8tion commented on microsvcs/gpt_generation/app/src/ext_api/init/topic.ts:
```diff
@@ -0,0 +1,3 @@
+const _topic_list = ["Diversity and Inclusion (D&I) Initiatives", "Cultural Competence", "Gender Equality", "LGBTQ+ Inclusion", "Disability Inclusion", "Racial and Ethnic Diversity", "Age Diversity", "Mental Health Awareness", "Workplace Inclusion", "Inclusive Technology", "Community Engagement"];
```
this file feels more like constants

cre8tion commented on microsvcs/gpt_generation/app/src/ext_api/init/user.ts:
```diff
@@ -0,0 +1,26 @@
+export async function _initialiseProfile(uid: string) {
+  const headers: Headers = new Headers()
+
+  headers.set('Content-Type', 'application/json')
```
You can extract out into a common headers if you need to reuse in the future

cre8tion commented on microsvcs/gpt_generation/app/src/ext_api/init/user.ts:
```diff
@@ -0,0 +1,26 @@
+
+  let result = await fetch(request)
+    .then(res => res.json())
+    .then(res => {
```
do you really need the last 'then'

cre8tion reviewed (commented):
This code is more complicated than the profile microservice, we will discuss more during the call if its possible

ZhiLinR commented:
fixes included:
1. refactored set headers method to its own file
2. middleware to handle standardized formatting for response and error
3. moved constants out of try block and removed controller error throwing

misc: 
1. changed some json parsing

cre8tion commented on microsvcs/gpt_generation/app/src/controller/profile_con.ts:
```diff
@@ -1,26 +1,23 @@
+        next({ success: true, status: 200, message: "Summarised", content: user_personality_summary });
     } catch (error) {
-        res.status(500).json(Handler.handleErrorResponse("Generation Failed"));
+        next({ success: true, status: 500, message: "Database Error" });
```
why is success true?

cre8tion commented on microsvcs/gpt_generation/app/src/controller/profile_con.ts:
```diff
@@ -15,9 +15,16 @@ profileRoute.get('/scenario/summary/:uid', async (req: Request, res: Response, n
+        if (user_personality_summary) {
+            next({ success: true, status: 200, message: "Summarised", content: user_personality_summary });
+        }else{
+            next({ success: true, status: 404, message: "Check User Input" });
```
success: false

cre8tion commented:
Make the success changes and LGTM. Feel free to merge afterwards

ZhiLinR commented:
omg 😭  thanks ill fix it
